### PR TITLE
Implement C++ unit test runner for Emscripten

### PR DESCRIPTION
**Pull request by emaxx-google**
This commit adds the support of building and running C++ unit tests in
Emscripten builds.

In order to achieve this, there's also a refactoring made that
simplifies how the tests' makefiles are organized and allows to abstract
away differences between Native Client and Emscripten.

This is an improvement for the Emscripten/WebAssembly build
infrastructure that is tracked by #177.

### CODE REVIEW

File: common/cpp_unit_test_runner/src/build_emscripten.mk [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	$(LIB_DIR)/libgtest.a \
+	$(LIB_DIR)/libgtest_main.a \
+
+$(GOOGLETEST_LIBS) &:
+	$(MAKE) -C $(ROOT_PATH)/third_party/googletest/webport/build
+
+# Documented at ../include.mk.
+#
+# Implementation notes:
+# The execution is performed via Node.js.
```
Comment (emaxx-google) [inline common/cpp_unit_test_runner/src/build_emscripten.mk:63]: @MahmoudAGawad PTAL at this part as well, in case you have any experience/suggestions here. Thanks.

File: common/cpp_unit_test_runner/src/build_emscripten.mk [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+# Flags passed to the Emscripten compiler/linker tools in test builds.
+#
+# Explanation:
+# MODULARIZE: Disable putting the Emscripten module JavaScript loading code into
+#   a factory function, so that the test runner module is loaded automatically
+#   on startup.
+TEST_ADDITIONAL_EMSCRIPTEN_FLAGS := \
+	-s MODULARIZE=0 \
+
+# Documented at ../include.mk.
```
Comment (Fabian-Sommer) [inline common/cpp_unit_test_runner/src/build_emscripten.mk:27]: nit: Documented in ../include.mk (same in other places in this CL)

File: common/cpp_unit_test_runner/src/build_emscripten.mk [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+# builds the C++ unit test runner using the Emscripten toolchain.
+
+# Flags passed to the Emscripten compiler/linker tools in test builds.
+#
+# Explanation:
+# MODULARIZE: Disable putting the Emscripten module JavaScript loading code into
+#   a factory function, so that the test runner module is loaded automatically
+#   on startup.
+TEST_ADDITIONAL_EMSCRIPTEN_FLAGS := \
+	-s MODULARIZE=0 \
```
Comment (Fabian-Sommer) [inline common/cpp_unit_test_runner/src/build_emscripten.mk:25]: Do we expect that all emscripten specific flags are applied to both compiling and linking? If not, it may be warranted to split this variable into two, or just use it directly in the two instances where it is used (though this solution does provide a nice place for the comment about MODULARIZE).

File: common/cpp/build/tests/Makefile [hunk 4]
```diff
@@ -21,46 +20,46 @@
...
 
-ADDITIONAL_TEST_CPPFLAGS := \
-	-I$(TEST_SOURCES_PATH) \
+CXXFLAGS := \
+	-I$(ROOT_SOURCES_PATH)/ \
+	-pedantic \
+	-Wall \
+	-Werror \
+	-Wextra \
+	-std=gnu++11 \
```
Comment (Fabian-Sommer) [inline common/cpp/build/tests/Makefile:56]: I would prefer `-std=c++11` (if that does not work, perhaps document the reason for it - not in this file, but in a more central place). Either thing should probably be a separate CL.

**Review by Fabian-Sommer — APPROVED** [review]
LGTM % nits

File: common/cpp/build/tests/Makefile (hunk 4 above)
Comment (emaxx-google) [reply common/cpp/build/tests/Makefile:53]: In NaCl there were some problems with `c++11`, which is why is had to use `gnu++11` (don't remember specific examples right now).
But you're right that it's not documented anywhere in this project, and also it'll probably be better to achieve "`gnu+11` in NaCl, and `c++11` in WebAssembly". I'll look into this later.

File: common/cpp_unit_test_runner/src/build_emscripten.mk (hunk 3 above)
Comment (emaxx-google) [reply common/cpp_unit_test_runner/src/build_emscripten.mk:25]: I'm not very experienced in Emscripten, but I was following their guideline that recommends using the same arguments: https://github.com/emscripten-core/emscripten/blob/9d37058cedb5ffa482202eb50812e67cd9bec810/src/settings.js#L30
I already experienced it with things like `-pthread` which result in weird compilation issues when it's missing somewhere. I can imagine some other flags can even cause runtime issues if used inconsistently.

Having said that, we definitely can move a flag into a specific location (like CXXFLAGS or LDFLAGS), as long as we 100% sure that it's needed only on a specific stage. But I haven't found any reference source for that yet.

File: common/cpp_unit_test_runner/src/build_emscripten.mk (hunk 1 above)
Comment (MahmoudAGawad) [reply common/cpp_unit_test_runner/src/build_emscripten.mk:63]: Unfortunately, I don't have experience with this :(

**Review by MahmoudAGawad — APPROVED** [review]
I defer to Fabian.

**Comment by emaxx-google** [comment]
Alright, thanks!
